src/devices/honeywell_cm921.c

```diff
@@ -0,0 +1,449 @@
+
+typedef struct {
+    uint8_t unknown_0; /* always == 0? */
+    uint8_t unknown_1; /* direction? */
+    uint8_t second;
+    uint8_t minute;
+    uint8_t hour : 5;
+    uint8_t day_of_week : 3;
+    uint8_t day;
+    uint8_t month;
+    uint8_t year[2];
+} X313F;
+
+
+data_t *add_hex_string(data_t *data, const char *name, const uint8_t *buffer, size_t buffer_length)
+{
+    if (buffer && buffer_length > 0)  {
+        char tstr[(buffer_length * 2) + 1];
+        char *p = tstr;
+        for (unsigned i = 0; i < buffer_length; i++, p+=2)
+            sprintf(p, "%02x", buffer[i]);
+        *p = '\0';
+        data = data_append(data, name, "", DATA_STRING, tstr, NULL);
+    }
+    return data;
+}
+
+typedef struct { uint8_t t; const char s[4]; } dev_map_entry_t;
+
+static const dev_map_entry_t device_map[] = {
+    { .t = 01, .s = "CTL" }, /* Controller */
+    { .t = 02, .s = "UFH" }, /* Underfloor heating (HCC80, HCE80) */
+    { .t = 03, .s = " 30" }, /* HCW82?? */
+    { .t = 04, .s = "TRV" }, /* Thermostatic radiator valve (HR80, HR91, HR92) */
+    { .t = 07, .s = "DHW" }, /* DHW sensor (CS92) */
+    { .t = 10, .s = "OTB" }, /* OpenTherm bridge (R8810) */
+    { .t = 12, .s = "THm" }, /* Thermostat with setpoint schedule control (DTS92E, CME921) */
+    { .t = 13, .s = "BDR" }, /* Wireless relay box (BDR91) (HC60NG too?) */
+    { .t = 17, .s = " 17" }, /* Dunno - Outside weather sensor? */
+    { .t = 18, .s = "HGI" }, /* Honeywell Gateway Interface (HGI80, HGS80) */
+    { .t = 22, .s = "THM" }, /* Thermostat with setpoint schedule control (DTS92E) */
+    { .t = 30, .s = "GWY" }, /* Gateway (e.g. RFG100?) */
+    { .t = 32, .s = "VNT" }, /* (HCE80) Ventilation (Nuaire VMS-23HB33, VMN-23LMH23) */
+    { .t = 34, .s = "STA" }, /* Thermostat (T87RF) */
+    { .t = 63, .s = "NUL" }, /* No device */
+};
+
+char* decode_device_id(const uint8_t device_id[3])
+{
+    uint8_t dev_type = device_id[0] >> 2;
+    const char *dev_name = " --";
+    for (size_t i = 0; i < sizeof(device_map) / sizeof(dev_map_entry_t); i++)
+        if (device_map[i].t == dev_type)
+            dev_name = device_map[i].s;
+
+    char *res = malloc(sizeof(char) * (6 + strlen(dev_name) + 1));
+    sprintf(res, "%3s:%06d", dev_name, (device_id[0] & 0x03) << 16 | (device_id[1] << 8) | device_id[2]);
+    return res;
+}
+
+data_t *decode_device_ids(const message_t *msg, data_t *data, int style) {
+    char *ds = calloc(11, msg->num_device_ids);
+
+    for (unsigned i = 0; i < msg->num_device_ids; i++) {
+        if (i != 0) strcat(ds, " ");
+
+        char *did = NULL;
+        if (style == 0)
+            did = decode_device_id(msg->device_id[i]);
+        else {
+            did = malloc(sizeof(char) * 7);
+            for (unsigned j = 0; j < 3; j++)
+                sprintf(did + 2*j, "%02x", msg->device_id[i][j]);
+        }
+        strcat(ds, did);
+        free(did);
+    }
+
+    return data_append(data, "Device IDs", "", DATA_STRING, ds, NULL);
+}
+
```

Comment:
All allocations here should be possible with static buffers.

Reply:
Point in case: it's leaking memory.

Reply:
That's been fixed too.